JDK-8312182: THPs cause huge RSS due to thread start timing issue

tstuefe opened the pull request:
If Transparent Huge Pages are unconditionally enabled (`/sys/kernel/mm/transparent_hugepage/enabled` contains `[always]`), Java applications that use many threads may see a huge Resident Set Size. That RSS is caused by thread stacks being mostly paged in. This page-in is caused by thread stack memory being transformed into huge pages by `khugepaged`; later, those huge pages usually shatter into small pages when Java guard pages are established at thread start, but the remaining splinter small pages remain paged in.

[JDK-8303215](https://bugs.openjdk.org/browse/JDK-8303215) attempted to fix this problem by making it unlikely that thread stack boundaries are aligned to THP page size. Unfortunately, that was not sufficient. We still see JVMs with huge footprints, especially if they did create many Java threads in rapid succession.

Note that this effect is independent of any JVM switches; in particular, it happens regardless of `-XX:+UseTransparentHugePages` or `-XX:+UseLargePages`.

Update: tests show that the interference of `khugepaged` also costs performance when starting threads, and this patch addresses both footprint and performance problems.

##### Demonstration:

Linux 5.15 on x64, glibc 2.31: 10000 idle threads with 100 MB pre-touched java heap, `-Xss2M`, on x64, will consume:

A) Baseline (THP disabled on system):  *369 MB*
B) THP="always", JDK-8303215 present: *1.5 GB .. >2 GB* (very wobbly)
C) THP="always", JDK-8303215 present, artificial delay after thread start: **20,6 GB** (!).


#### Cause:

The problem is caused by timing. When we create multiple Java threads, the following sequence of actions happens:

In the parent thread:
- the parent thread calls `pthread_create(3)`
- `pthread_create(3)` creates the thread stack by calling `mmap(2)`
- `pthread_create(3)` calls `clone(2)` to start the child thread
- repeat to start more threads

Each child thread:
- queries its stack dimensions
- handshakes with the parent to signal lifeness
- establishes guard pages at the low end of the stack

The thread stack mapping is established in the parent thread; the guard pages are placed by the child threads. There is a time window in which the thread stack is already mapped into address space, but guard pages still need to be placed.

If the parent is faster than the children, it will have created mappings faster than the children can place guard pages on them.

For the kernel, these thread stacks are just anonymous mappings. It places them adjacent to each other to reduce address space fragmentation. As long as no guard pages are placed yet, all these thread stack mappings (VMAs) have the same attributes - same permission bits, all anonymous. Hence, the kernel will fold them into a single large VMA. 

That VMA may be large enough to be eligible for huge pages. Now the JVM races with the `khugepaged`: If `khugepaged` is faster than the JVM, it will have converted that larger VMA partly or fully into hugepages before the child threads start creating guard pages:


```
glibc does              Kernel folds them        khugepaged creates
1 mmap/stack:           into one VMA (still      huge pages (paged in):
                        not paged in):
+---------+             +---------+              +---------+ 
|         |             |         |              |         | 
| stack A |             |         |              | hugepage|
|         |             |         |              | ------- | 
+---------+             |         |              |         | 
|         |             |         |              | hugepage| 
| stack B |    ====>    |  VMA    |    ====>     | ------- | 
|         |             |         |              |         | 
+---------+             |         |              | hugepage| 
|         |             |         |              | ------- | 
| stack C |             |         |              |         | 
|         |             |         |              | hugepage| 
+---------+             +---------+              +---------+
```

The child threads will catch up and create guard pages. That will splinter the large VMA into several smaller VMAs (two for each thread, one for the usable thread section, and one protected for the guards). Each of these VMAs will typically be smaller than a huge page, and typically not huge-page-aligned. The huge pages created by `khugepaged` will mostly shatter into small pages, but these small pages remain paged-in. Effect: we pay memory for the whole thread stacks even though the threads did not start yet.

This is a similar effect as described in [JDK-8303215](https://bugs.openjdk.org/browse/JDK-8303215); but we assumed it only affects individual threads when it affects whole regions of adjacent thread stacks.

#### Example:

Let's create three threads. Each thread stack, including guard pages, is 2M + 4K sized (+4K because of JDK-8303215).

Their thread stacks will be located at: ( `[base .. end .. guard]`:

```
T1: [7feea53ff000 .. 7feea5202000 .. 7feea51fe000] 
T2: [7feea5600000 .. 7feea5403000 .. 7feea53ff000] 
T3: [7feea5801000 .. 7feea5604000 .. 7feea5600000]
```

After `pthread_create(3)`, their thread stacks exist without JVM guard pages. Kernel merges the VMAs of their thread stacks into a single mapping > 6MB. `khugepaged` then coalesces their small pages into 3 huge pages:

```
7feea51fe000-7feea5801000 rw-p 00000000 00:00 0    <<<------- all three stacks as one VMA
Size:               6156 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:                6148 kB
Pss:                6148 kB
Shared_Clean:          0 kB
Shared_Dirty:          0 kB
Private_Clean:         0 kB
Private_Dirty:      6148 kB
Referenced:         6148 kB
Anonymous:          6148 kB
LazyFree:              0 kB
AnonHugePages:      6144 kB      <<<---------- 3x2MB huge pages
ShmemPmdMapped:        0 kB
FilePmdMapped:         0 kB
Shared_Hugetlb:        0 kB
Private_Hugetlb:       0 kB
Swap:                  0 kB
SwapPss:               0 kB
Locked:                0 kB
THPeligible:    1    
VmFlags: rd wr mr mw me ac sd 
```

Threads start and create their respective guard pages. The single VMA splinters into 6 smaller VMAs. The huge pages shatter into small pages that remain paged-in:

```
7feea51fe000-7feea5202000 ---p 00000000 00:00 0   <<----- guard pages for T1
Size:                 16 kB
...
7feea5202000-7feea53ff000 rw-p 00000000 00:00 0   <<------ thread stack for T1
Size:               2036 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:                2036 kB
Pss:                2036 kB
Private_Dirty:      2036 kB   <<<--------  all pages resident
...
7feea53ff000-7feea5403000 ---p 00000000 00:00 0   <<----- guard pages for T2
Size:                 16 kB
...
7feea5403000-7feea5600000 rw-p 00000000 00:00 0   <<------ thread stack for T2
Size:               2036 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:                2036 kB
Pss:                2036 kB
Private_Dirty:      2036 kB   <<<--------  all pages resident
...
7feea5600000-7feea5604000 ---p 00000000 00:00 0    <<----- guard pages for T3 
Size:                 16 kB
...
7feea5604000-7feea5801000 rw-p 00000000 00:00 0    <<------ thread stack for T3
Size:               2036 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:                2036 kB
Pss:                2036 kB
Private_Dirty:      2036 kB   <<<--------  all pages resident
...
```

#### Fix

The mitigation for the huge RSS buildup involves letting the glibc allocate guards for Java thread stacks. We usually avoid that since Java thread stacks are already guarded by JVM guard pages. However, here we use the fact that the glibc guard page is established by `pthread_create(3)` **before** starting the child thread. Therefore adjacent thread stack VMAs will be separated by a guard page - we close the time window within which the kernel sees a large VMA. 

The fix works: the aforementioned 20.6 GB RSS example, with the fix applied, drops back to ~400 MB RSS.

As a byproduct, this patch fixes [JDK-8310687](https://bugs.openjdk.org/browse/JDK-8310687): the mitigation will only be activated if THPs are enabled unconditionally on the system (THP mode = "always"). And we use the correct page size for THPs.

Finally, note that this solution has a slight disadvantage: threads now have two guards, the JVM guard pages and the glibc guard page. Currently, the kernel cannot merge these VMAs because their attributes differ. This problem may be solvable; I opened [JDK-8312211](https://bugs.openjdk.org/browse/JDK-8312211)  to track this problem.

/issue JDK-8310687

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8312182](https://bugs.openjdk.org/browse/JDK-8312182): THPs cause huge RSS due to thread start timing issue (**Bug** - P3)
 * [JDK-8310687](https://bugs.openjdk.org/browse/JDK-8310687): JDK-8303215 is incomplete (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Poonam Bajaj](https://openjdk.org/census#poonam) (@poonamparhar - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14919/head:pull/14919` \
`$ git checkout pull/14919`

Update a local copy of the PR: \
`$ git checkout pull/14919` \
`$ git pull https://git.openjdk.org/jdk.git pull/14919/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14919`

View PR using the GUI difftool: \
`$ git pr show -t 14919`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14919.diff">https://git.openjdk.org/jdk/pull/14919.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14919#issuecomment-1640486548)

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
+  // We prevent that by letting the glibc allocate a guard page, which causes a VMA with different
+  // permission bits to separate two ajacent thread stacks and therefore prevent merging stacks
+  // into one VMA.
+  if (PreventTHPsForThreadStacks) {
+    guard_size = MAX2(guard_size, os::vm_page_size());
```
I think this would be better hidden in `os::Linux::default_guard_size` and only applied to JavaThreads.

tstuefe replied:
> I think this would be better hidden in os::Linux::default_guard_size

I don't like this, since it would keep those two mitigations we apply - adding the glibc guard page and altering the stack size to be THP unaligned - separate in code. I find it clearer to have them in one place.

> and only applied to JavaThreads.

No, it makes sense to apply this to any thread that does not by-default run with a glibc guard page. But in practice, it is as you wish it to be since only Java Threads run without glibc guard pages.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
+  // into one VMA.
+  if (PreventTHPsForThreadStacks) {
+    guard_size = MAX2(guard_size, os::vm_page_size());
+    // Add an additional page to the stack size to reduce its chances of getting huge page aligned
+    // so that the stack does not get backed by a transparent huge page.
```
I don't think these two adjustments should be combined like this. The existing +1 may be sufficient for some folk who do not want the cost of creating a glibc guard page as well.

tstuefe replied:
I disagree, see earlier remarks. "May be sufficient" is vague - either you want to prevent THPs forming in thread stacks, or you are okay with it. There is no point in doing one without the other.

If you want to prevent them, you have to do both since the effects laid out in this PR are not predictable. Even if your app does not create - and why would you care for that one guard page then - the JVM creates a bunch of threads in quick succession too and thus suffers from the same effects.

And if you want to have THPs, then you should disable both mitigations together.

dholmes-ora reviewed (changes requested):
Wow what a mess! The OS is really working against us in this area. We should not have to be playing these silly games.

There is a caution in `os::Linux::default_guard_size` that glibc guard pages are expensive to create so what are we paying for this?

Thanks.

tstuefe commented:
> Wow what a mess! The OS is really working against us in this area. We should not have to be playing these silly games.
> 
> There is a caution in `os::Linux::default_guard_size` that glibc guard pages are expensive to create so what are we paying for this?

Less, actually :-) The patch is a performance win too.

Code-wise, glibc seems to do one more `mmap(2)` call if we use guard pages, at least for the standard code paths. 

I wanted to see if that matters, so I tested with THP mode set to "always" on my 48-core Linux 5.15 glibc 2.31 box. I tested with two stack sizes (1 MB, 2 MB), with and without patch. The test measures the time it takes to spawn 10000 threads.

| Stack size (kb)|  with patch (ms) | without patch (ms) | Delta |
|----------------|------------------|--------------------|-------|
| 1024           |  723 +- 11       |   807 +- 18        | +12%  |
| 2048           |  710 +- 13       |  1621 +- 91        | +128% |

The numbers are consistently better with my patch. When the stack size approaches THP page size, the difference gets *huge*!

This makes sense since `khugepaged` transforms thread pages under our noses while child threads, freshly started, try to use the same pages. So we get a lot of interference. The closer the stack size is to THP size, the more likely this is. And we use more memory too, which costs bandwidth.

I repeated the test with THPs disabled, and sure enough, the difference vanished.

Bottomline, the cost of the additional mmap is dwarfed by the win of not having to fight khugepaged for the thread stacks.

> I don't think these two adjustments should be combined like this. The existing +1 may be sufficient for some folk who do not want the cost of creating a glibc guard page as well.

I disagree; I'd like to enable this patch by default as a whole if we run with THP=always; we save performance and memory. There are not many downsides.

poonamparhar commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
+    // Add an additional page to the stack size to reduce its chances of getting huge page aligned
+    // so that the stack does not get backed by a transparent huge page.
+    if (HugePages::thp_pagesize() > 0 &&
+        is_aligned(stack_size, HugePages::thp_pagesize())) {
```
Do we want to add this additional page even when the stack_size is less than the thp_pagesize?

tstuefe replied:
The aligned test should suffice: if stack size is aligned to THP page size, the smallest non-zero value is that pagesize. It cannot be 0, we would have noticed by now.

poonamparhar replied:
Yes, makes sense.

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
+  // permission bits to separate two ajacent thread stacks and therefore prevent merging stacks
+  // into one VMA.
+  if (PreventTHPsForThreadStacks) {
+    guard_size = MAX2(guard_size, os::vm_page_size());
```
Hmm. We already have a comment in place in `os::Linux::default_guard_size(thr_type);`. Maybe @dholmes-ora is right; I will add this logic to default_guard_size and extent the comment over there.

dholmes-ora replied:
Thanks for moving the logic.

poonamparhar replied:
Thanks for making this change!

tstuefe commented:
@poonamparhar @dholmes-ora Thanks for your respective feedback. New version slightly moves the guard page calculation around and improves comment, functionally the same.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,19 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
+
+  if (PreventTHPsForThreadStacks) {
+    // In addition to the glibc guard page that prevents inter-thread-stack hugepage
+    // coalescation (see comment in os::Linux::default_guard_size()), we also make
```
s/coalescation/coalescing/

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3077,6 +3085,27 @@ bool os::Linux::libnuma_init() {
+    // into one VMA.
+    //
+    // Yes, this means we have two guard sections - the glibc and the JVM one - per thread. But the
+    // cost for that one extra protected page is dwarfed a large win in performance and memory that
```
s/dwarfed a/dwarfed by a/

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3077,6 +3085,27 @@ bool os::Linux::libnuma_init() {
+    //
+    // Yes, this means we have two guard sections - the glibc and the JVM one - per thread. But the
+    // cost for that one extra protected page is dwarfed a large win in performance and memory that
+    // avoiding interference by khugepaged buys us.
```
s/by/from/

dholmes-ora commented on src/hotspot/os/linux/globals_linux.hpp:
```diff
@@ -89,6 +89,13 @@
           "to disable both the override and the printouts."             \
           "See prctl(PR_SET_TIMERSLACK) for more info.")                \
                                                                         \
+  product(bool, PreventTHPsForThreadStacks, true, EXPERIMENTAL,         \
+          "If true, the JVM will attempt to prevent formation of "      \
+          "transparent huge pages in thread stacks.")                   \
```
Can we expand this to state that it can be turned off by ergonomics. Or better yet perhaps this should be off by default and potentially turned on by ergonomics?

tstuefe replied:
As in, if THP=always, this will be turned on unless explicitly turned off? I dislike that I cannot see the default value in globals.hpp (for numerical args, "0" is always an indication for "no value", but for bool, it looks like the real default value).

What do you think about "AllowtTHPInThreadStacks" - default off, can be switched on manually to disable the mitigation if someone really wants THP for his thread stacks?

Alternative name "DisableTHPThreadStackMitigation" or somesuch, but I don't like the negation in the name.

dholmes-ora replied:
The value in globals.hpp cannot tell you what the "default" is because ergonomics may change it no matter which way we decide to express this. For a user that knows nothing about THP mode (e.g. me!) I can't tell whether this will be on or off or whether I may need to turn it off or on.

If the intent is that we normally enable the mitigation when THP mode = always, but the user can turn it off then something like `DisableTHPThreadStackMitigation` seems more appropriate to me.

tstuefe commented:
New version:
- changed flag name from `PreventTHPsForThreadStacks` -> `DisableTHPStackMitigation` (inverse logic) and made it diagnostic
- comment corrections
- changed the negative tests to be a jtreg / manual test (@dholmes-ora is that the right way to go here? I want a test that does not run automatically, but that can be run from command line)

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -911,6 +915,19 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
   size_t stack_size = os::Posix::get_initial_stack_size(thr_type, req_stack_size);
   size_t guard_size = os::Linux::default_guard_size(thr_type);
+
+  if (!DisableTHPStackMitigation) {
```
Good catch.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3740,6 +3769,21 @@ void os::large_page_init() {
+  // is usually unwanted for thread stacks. Therefore we attempt to prevent THP formation in
+  // thread stacks unless the user explicitly allowed THP formation by manually disabling
+  // -XX:+DisableTHPStackMitigation.
+  if (HugePages::thp_mode() == THPMode::always) {
```
What controls this? The user launching the VM? The OS configuration? Or both?

tstuefe replied:
The OS, so ultimately the Admin that set up the OS.

dholmes-ora reviewed (commented):
This seems reasonable to me now - not that I understand THP any better. I will run the test through our CI for good measure.

Thanks.

dholmes-ora reviewed (approved):
Test passes in our CI.

tstuefe commented:
> Test passes in our CI.

Many thanks, David.

poonamparhar reviewed (approved):
The changes look good to me.

tstuefe commented:
Thanks @poonamparhar, for the review and the analysis of JDK-8303215!

/integrate
